### PR TITLE
Fix "Cannot read property 'split' of undefined" for interconnected file based dependencies

### PR DESCRIPTION
**Pull request by jjudd**
**Summary**
Currently, when you have interconnected file dependencies, yarn install fails after a lockfile has been generated. This failure occurs because yarn calls `split` on the package's resolved url, which is `undefined`. This fix makes sure we have a resolved url before trying to call `split` on it.

This is described in #2256. You can find a repo that reproduces this issue at https://github.com/mstijak/yarn-file-dep-repro

As described in #2256, running `yarn install` on the above repo produces the following lockfile:

```
# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"a@file:./packages/a":
  version "1.0.2"
  dependencies:
    b "1.0.0"

b@*:
  version "1.0.0"

"b@file:./packages/b":
  version "1.0.0"
```

When you run `yarn install` a second, third, fourth, etc. time, yarn fails when trying to resolve the following package:

```
b@*:
  version "1.0.0"
```

It attempts to use a tarball resolver. In doing so, it calls split on a url that is undefined. My fix just checks makes sure the url is defined before trying to explode it. I'm not terribly familiar with yarn's code, so I'm not sure if this is a good fix or not.

**Test plan**
*To reproduce the bug*
1. Check out https://github.com/mstijak/yarn-file-dep-repro
2. Run `yarn install`. Notice that `yarn.lock` now exists.
3. Run `yarn install` again. This should fail with the following error
```
Trace: 
  TypeError: Cannot read property 'split' of undefined
      at Object.explodeHashedUrl (/home/james/yarn/lib/util/version.js:8:20)
      at PackageRequest.getLocked (/home/james/yarn/lib/package-request.js:128:59)
      at /home/james/yarn/lib/resolvers/registries/npm-resolver.js:202:37
      at next (native)
      at step (/home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
      at /home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:35:14
      at Promise.F (/home/james/yarn/node_modules/core-js/library/modules/_export.js:35:28)
      at /home/james/yarn/node_modules/babel-runtime/helpers/asyncToGenerator.js:14:12
      at NpmResolver.resolve (/home/james/yarn/lib/resolvers/registries/npm-resolver.js:237:7)
      at /home/james/yarn/lib/package-request.js:183:23
```
*To verify the fix*
With the fix applied repeat the steps above. On step 3, `yarn install` should succeed without error.

**Gifs demonstrating the fix follow**
*Broken*
broken

*Fixed*
fixed

### CODE REVIEW

**Comment by bestander** [comment]
Thanks for the fix, @jjudd.
Would you add a test please for this scenario?
Something similar to this https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/install/integration.js#L215

**Comment by jjudd** [comment]
Sure thing. Working on it now.

**Comment by jjudd** [comment]
I added the test @bestander. It fails without the fix and succeeds with it. Please let me know if you would like to do anything else.

File: __tests__/commands/install/integration.js [hunk 1]
```diff
@@ -236,6 +236,13 @@ test.concurrent('install file: protocol without cache', async (): Promise<void>
   });
 });
 
+test.concurrent('install file: local packages with local dependencies', async (): Promise<void> => {
+  await runInstall({}, 'install-file-local-dependency', async (config, reporter) => {
+    const reinstall = new Install({}, config, reporter, await Lockfile.fromDirectory(config.cwd));
+    await reinstall.init();
```
Comment (bestander) [inline __tests__/commands/install/integration.js]: Would be great to have an assert in the test that you have the expected state.

**Comment by bestander** [comment]
Just a nit and ready to merge, thanks

File: __tests__/commands/install/integration.js (hunk 1 above)
Comment (jjudd) [inline __tests__/commands/install/integration.js]: What would the expected state be here? I'm fine with adding that, I just want to make sure I'm asserting the right thing. At the moment, I'm relying on the exception being thrown to fail the test. I could catch the exception and set a boolean and assert on that boolean or something similar.

Comment (bestander) [inline __tests__/commands/install/integration.js]: It is either assert that exception not thrown or assert that module has been installed in node modules

Comment (jjudd) [inline __tests__/commands/install/integration.js]: Makes sense. Thanks for the feedback. I updated the test to assert that the module is installed in node modules.

**Comment by bestander** [comment]
Amazing PR, thanks @jjudd .
The tests, the description with the gif - great

**Comment by jjudd** [comment]
Thanks for the kind words. Glad I could help out :)

**Pull request merged**